sycl/source/detail/queue_impl.hpp

```diff
@@ -118,7 +118,7 @@ class queue_impl {
   }
 
   ~queue_impl() {
-    throw_asynchronous();
+    wait_and_throw();
```

Comment:
It would be good to add a comment explaining it

Reply:
Is such explanation sufficient or you want something else?

Reply:
It would be good to give an example where the wait is crucial for correctness, and also mention what SYCL spec says about this.

Reply:
> This behavior contradicts SYCL specification.

What exactly in the spec?

Reply:
It says that `queue` destructor is not blocking

Reply:
So, why are we adding something that contradicts the spec?

Reply:
There is one more "problem" - the `queue_impl` is not guaranteed to be destroyed when all user visible `sycl::queue` objects are released because we reference `queue_impl` internally.

Reply:
I don't understand what problem this is supposed to be fixing.  The idea behind having the queue destructor be non-blocking is that destroying the queue shouldn't necessarily actually get rid of the allocation.  "destroying" the queue should really just decrement it's reference counter.  And if it's reference counter becomes 0 then the queue can truly be deleted.  I suspect that all the places that need to increment or decrement the ref counter are not really doing so.

Reply:
> And if it's reference counter becomes 0 then the queue can truly be deleted.

What does "deleted" mean? If application is exiting, and there're still running kernels, what should happen to those kernels? Abort the execution? Wait for all kernels to finish? Who's accountable for this? User? SYCL runtime? It's a lot of questions. Putting `wait()` to `~queue_impl()` seems to me like an easy way out of this uncertainty.

Another way we could take is waiting for all tasks to finish in `Scheduler` destructor. But I don't see how it can be easily done with current `Scheduler`.

> I suspect that all the places that need to increment or decrement the ref counter are not really doing so.

`queue_impl()` is always wrapped with `std::shared_ptr`, and I'm not aware of any memory leaks in Scheduler, so, it should be fine.

Reply:
You ask: If the application is exiting, what should happen to kernels that are still running??
My question: If the application is exiting, why are there kernels that are still running?  Seems like an application error to me, and that doing this change would penalize correct applications in trying to fix broken applications.

Reply:
@kbsmith-intel sounds reasonable. Do you agree then, that this example is incorrect?

```c++
 int main() {
     queue Q;
     Q.submit([&](handler &h){
        stream out(1024, 256, h);
        h.parallel_for(range{8}, [=](id<1> idx){
          out << idx << ")\n";
        });
      });
     return 0;
 }
```

By the time we hit `return 0`, the kernel is still running.

Also tagging @Pennycook, @mkinsner, @jbrodman, @keryell. I'd appreciate your inputs.

Reply:
Yes, I think this example is a broken program. But you have tagged the right people to help us make sure that it is an incorrect program.

Reply:
> By the time we hit `return 0`, the kernel is still running.

A possible variation: by the time we hit `return 0`, the kernel has not started yet.

Reply:
> You ask: If the application is exiting, what should happen to kernels that are still running??
> My question: If the application is exiting, why are there kernels that are still running? Seems like an application error to me, and that doing this change would penalize correct applications in trying to fix broken applications.

I disagree. A kernel that is still running is a valid use case in FPGA programming, for example reading input from I/O pipe inside of an infinite loop.

Reply:
I agree this is a useful example but in that case it makes sense to have the host waiting for the kernel otherwise the FPGA might be wiped out for another application.